File: rust/src/winenv/defs/windows/com.rs
use crate::r#struct::{EmuStruct, Ptr};
use uuid::Uuid;

pub const S_OK: u32 = 0;
pub const RPC_C_AUTHN_LEVEL_DEFAULT: u32 = 0;
pub const RPC_C_AUTHN_LEVEL_NONE: u32 = 1;
pub const RPC_C_AUTHN_LEVEL_CONNECT: u32 = 2;
pub const RPC_C_AUTHN_LEVEL_CALL: u32 = 3;
pub const RPC_C_AUTHN_LEVEL_PKT: u32 = 4;
pub const RPC_C_AUTHN_LEVEL_PKT_INTEGRITY: u32 = 5;
pub const RPC_C_AUTHN_LEVEL_PKT_PRIVACY: u32 = 6;

pub const RPC_C_IMP_LEVEL_DEFAULT: u32 = 0;
pub const RPC_C_IMP_LEVEL_ANONYMOUS: u32 = 1;
pub const RPC_C_IMP_LEVEL_IDENTIFY: u32 = 2;
pub const RPC_C_IMP_LEVEL_IMPERSONATE: u32 = 3;
pub const RPC_C_IMP_LEVEL_DELEGATE: u32 = 4;

pub const CLSID_WBEM_LOCATOR: &str = "{4590F811-1D3A-11D0-891F-00AA004B2E24}";
pub const CLSID_IWBEM_CONTEXT: &str = "{674B6698-EE92-11D0-AD71-00C04FD8FDFF}";
pub const IID_IWBEM_LOCATOR: &str = "{DC12A687-737F-11CF-884D-00AA004B2E24}";
pub const IID_IWBEM_CONTEXT: &str = "{44ACA674-E8FC-11D0-A07C-00C04FB68820}";

#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct IUnknown {
    pub QueryInterface: Ptr,
    pub AddRef: Ptr,
    pub Release: Ptr,
}
impl EmuStruct for IUnknown {}

#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct IWbemLocator {
    pub IUnknown: IUnknown,
    pub ConnectServer: Ptr,
}
impl EmuStruct for IWbemLocator {}

#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct IMalloc {
    pub IUnknown: IUnknown,
    pub Alloc: Ptr,
    pub Realloc: Ptr,
    pub Free: Ptr,
    pub GetSize: Ptr,
    pub DidAlloc: Ptr,
    pub HeapMinimize: Ptr,
}
impl EmuStruct for IMalloc {}

#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct IWbemServices {
    pub IUnknown: IUnknown,
    pub OpenNamespace: Ptr,
    pub CancelAsyncCall: Ptr,
    pub QueryObjectSink: Ptr,
    pub GetObject: Ptr,
    pub GetObjectAsync: Ptr,
    pub PutClass: Ptr,
    pub PutClassAsync: Ptr,
    pub DeleteClass: Ptr,
    pub DeleteClassAsync: Ptr,
    pub CreateClassEnum: Ptr,
    pub CreateClassEnumAsync: Ptr,
    pub PutInstance: Ptr,
    pub PutInstanceAsync: Ptr,
    pub DeleteInstance: Ptr,
    pub DeleteInstanceAsync: Ptr,
    pub CreateInstanceEnum: Ptr,
    pub CreateInstanceEnumAsync: Ptr,
    pub ExecQuery: Ptr,
    pub ExecQueryAsync: Ptr,
    pub ExecNotificationQuery: Ptr,
    pub ExecNotificationQueryAsync: Ptr,
    pub ExecMethod: Ptr,
    pub ExecMethodAsync: Ptr,
}
impl EmuStruct for IWbemServices {}

#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct IWbemContext {
    pub IUnknown: IUnknown,
    pub Clone: Ptr,
    pub GetNames: Ptr,
    pub BeginEnumeration: Ptr,
    pub Next: Ptr,
    pub EndEnumeration: Ptr,
    pub SetValue: Ptr,
    pub GetValue: Ptr,
    pub DeleteValue: Ptr,
    pub DeleteAll: Ptr,
}
impl EmuStruct for IWbemContext {}

#[derive(Clone, Copy, Debug)]
pub struct ComInterface<T: EmuStruct + Copy> {
    pub iface: T,
    pub address: u64,
    pub name: &'static str,
}

impl<T: EmuStruct + Copy> ComInterface<T> {
    pub fn new(iface: T, name: &'static str) -> Self {
        Self {
            iface,
            address: 0,
            name,
        }
    }
}

pub fn get_define_int(define: u32, prefix: &str) -> Option<&'static str> {
    let defs = [
        ("RPC_C_AUTHN_LEVEL_DEFAULT", RPC_C_AUTHN_LEVEL_DEFAULT),
        ("RPC_C_AUTHN_LEVEL_NONE", RPC_C_AUTHN_LEVEL_NONE),
        ("RPC_C_AUTHN_LEVEL_CONNECT", RPC_C_AUTHN_LEVEL_CONNECT),
        ("RPC_C_AUTHN_LEVEL_CALL", RPC_C_AUTHN_LEVEL_CALL),
        ("RPC_C_AUTHN_LEVEL_PKT", RPC_C_AUTHN_LEVEL_PKT),
        ("RPC_C_AUTHN_LEVEL_PKT_INTEGRITY", RPC_C_AUTHN_LEVEL_PKT_INTEGRITY),
        ("RPC_C_AUTHN_LEVEL_PKT_PRIVACY", RPC_C_AUTHN_LEVEL_PKT_PRIVACY),
        ("RPC_C_IMP_LEVEL_DEFAULT", RPC_C_IMP_LEVEL_DEFAULT),
        ("RPC_C_IMP_LEVEL_ANONYMOUS", RPC_C_IMP_LEVEL_ANONYMOUS),
        ("RPC_C_IMP_LEVEL_IDENTIFY", RPC_C_IMP_LEVEL_IDENTIFY),
        ("RPC_C_IMP_LEVEL_IMPERSONATE", RPC_C_IMP_LEVEL_IMPERSONATE),
        ("RPC_C_IMP_LEVEL_DELEGATE", RPC_C_IMP_LEVEL_DELEGATE),
    ];
    defs.iter()
        .find_map(|(name, value)| (name.starts_with(prefix) && *value == define).then_some(*name))
}

pub fn get_define_str(define: &str, prefix: &str) -> Option<&'static str> {
    let defs = [
        ("CLSID_WBEM_LOCATOR", CLSID_WBEM_LOCATOR),
        ("CLSID_IWBEM_CONTEXT", CLSID_IWBEM_CONTEXT),
        ("IID_IWBEM_LOCATOR", IID_IWBEM_LOCATOR),
        ("IID_IWBEM_CONTEXT", IID_IWBEM_CONTEXT),
    ];
    defs.iter()
        .find_map(|(name, value)| (name.starts_with(prefix) && *value == define).then_some(*name))
}

pub fn get_clsid(define: &str) -> Option<&'static str> {
    get_define_str(define, "CLSID_")
}

pub fn get_iid(define: &str) -> Option<&'static str> {
    get_define_str(define, "IID_")
}

pub fn get_rpc_authlevel(define: u32) -> Option<&'static str> {
    get_define_int(define, "RPC_C_AUTHN_LEVEL_")
}

pub fn get_rcp_implevel(define: u32) -> Option<&'static str> {
    get_define_int(define, "RPC_C_IMP_LEVEL_")
}

pub const IFACE_TYPES: &[&str] = &[
    "IUnknown",
    "IMalloc",
    "IWbemLocator",
    "IWbemServices",
    "IWbemContext",
];

pub fn convert_guid_bytes_to_str(guid_bytes: [u8; 16]) -> String {
    format!("{{{}}}", Uuid::from_bytes_le(guid_bytes).hyphenated()).to_uppercase()
}

File: secpp/windows/com.cpp
// com.cpp
#include "com.h"
#include "winemu.h"
#include <stdexcept>
#include <cstring>

//  COM interface field definitions 

namespace comdefs {

// IUnknown: 3 method pointers
const std::vector<ComField> IUnknown_fields = {
    {"QueryInterface", false},
    {"AddRef",         false},
    {"Release",        false},
};

// IMalloc: IUnknown (inherited) + 6 methods
const std::vector<ComField> IMalloc_fields = {
    {"IUnknown",    true},   // inherited interface
    {"Alloc",       false},
    {"Realloc",     false},
    {"Free",        false},
    {"GetSize",     false},
    {"DidAlloc",    false},
    {"HeapMinimize", false},
};

// IWbemLocator: IUnknown + 1 method
const std::vector<ComField> IWbemLocator_fields = {
    {"IUnknown",      true},
    {"ConnectServer", false},
};

// IWbemServices: IUnknown + 23 methods
const std::vector<ComField> IWbemServices_fields = {
    {"IUnknown",             true},
    {"OpenNamespace",        false},
    {"CancelAsyncCall",      false},
    {"QueryObjectSink",      false},
    {"GetObject",            false},
    {"GetObjectAsync",       false},
    {"PutClass",             false},
    {"PutClassAsync",        false},
    {"DeleteClass",          false},
    {"DeleteClassAsync",     false},
    {"CreateClassEnum",      false},
    {"CreateClassEnumAsync", false},
    {"PutInstance",          false},
    {"PutInstanceAsync",     false},
    {"DeleteInstance",       false},
    {"DeleteInstanceAsync",  false},
    {"CreateInstanceEnum",   false},
    {"CreateInstanceEnumAsync", false},
    {"ExecQuery",            false},
    {"ExecQueryAsync",       false},
    {"ExecNotificationQuery",      false},
    {"ExecNotificationQueryAsync", false},
    {"ExecMethod",           false},
    {"ExecMethodAsync",      false},
};

// IWbemContext: IUnknown + 9 methods
const std::vector<ComField> IWbemContext_fields = {
    {"IUnknown",        true},
    {"Clone",           false},
    {"GetNames",        false},
    {"BeginEnumeration", false},
    {"Next",            false},
    {"EndEnumeration",  false},
    {"SetValue",        false},
    {"GetValue",        false},
    {"DeleteValue",     false},
    {"DeleteAll",       false},
};

// IFACE_TYPES map: name -> fields
const std::map<std::string, std::vector<ComField>> IFACE_TYPES = {
    {"IUnknown",      IUnknown_fields},
    {"IMalloc",       IMalloc_fields},
    {"IWbemLocator",  IWbemLocator_fields},
    {"IWbemServices", IWbemServices_fields},
    {"IWbemContext",  IWbemContext_fields},
};

} // namespace comdefs

//  COM implementation 

COM::COM(const speakeasy::SpeakeasyConfig& cfg)
    : config(cfg) {
}

std::shared_ptr<ComInterface> COM::get_interface(void* emu, size_t ptr_size, const std::string& name) {
    // Check cache
    auto it = interfaces.find(name);
    if (it != interfaces.end()) {
        return it->second;
    }

    // Look up the interface type
    auto iface_it = comdefs::IFACE_TYPES.find(name);
    if (iface_it == comdefs::IFACE_TYPES.end()) {
        throw Win32EmuError("Invalid COM interface: " + name);
    }

    const auto& fields = iface_it->second;
    auto ci = std::make_shared<ComInterface>(fields, name, ptr_size);
    auto* winemu = static_cast<WindowsEmulator*>(emu);

    // Allocate vtable memory
    uint64_t com_ptr = winemu->mem_map(ci->iface_size, std::nullopt, 0x7, "emu.COM." + name);
    ci->address = com_ptr;

    // Build the vtable
    size_t field_offset = 0;
    for (const auto& field : fields) {
        if (field.is_interface) {
            // Inherited interface (e.g. IUnknown inside IMalloc)
            // Look up the nested interface's fields
            auto nested_it = comdefs::IFACE_TYPES.find(field.name);
            if (nested_it == comdefs::IFACE_TYPES.end()) {
                throw Win32EmuError("COM interface " + name +
                                    " inherits unsupported interface " + field.name);
            }

            const auto& nested_fields = nested_it->second;
            for (const auto& subfield : nested_fields) {
                // Nested interface fields should be method pointers
                if (!subfield.is_interface) {
                    std::string method_name = field.name + "_" + subfield.name;
                    uint64_t addr = winemu->add_callback("com_api", method_name);
                    std::vector<uint8_t> addr_bytes(ptr_size, 0);
                    for (size_t i = 0; i < ptr_size; i++) {
                        addr_bytes[i] = static_cast<uint8_t>((addr >> (i * 8)) & 0xFF);
                    }
                    winemu->mem_write(com_ptr + field_offset, addr_bytes);
                }
                field_offset += ptr_size;
            }
        } else {
            // Direct method pointer
            std::string method_name = name + "_" + field.name;
            uint64_t addr = winemu->add_callback("com_api", method_name);
            std::vector<uint8_t> addr_bytes(ptr_size, 0);
            for (size_t i = 0; i < ptr_size; i++) {
                addr_bytes[i] = static_cast<uint8_t>((addr >> (i * 8)) & 0xFF);
            }
            winemu->mem_write(com_ptr + field_offset, addr_bytes);
            field_offset += ptr_size;
        }
    }

    // Cache and return
    interfaces[name] = ci;
    return ci;
}
